L1Trigger/L1THGCal/src/backend/HGCalMulticlusteringHistoImpl.cc

```diff
@@ -0,0 +1,358 @@
+#include "L1Trigger/L1THGCal/interface/backend/HGCalShowerShape.h"
+#include "DataFormats/Math/interface/deltaR.h"
+#include "DataFormats/Math/interface/deltaPhi.h"
+
+HGCalMulticlusteringHistoImpl::HGCalMulticlusteringHistoImpl( const edm::ParameterSet& conf ) :
+    dr_(conf.getParameter<double>("dR_multicluster")),
+    ptC3dThreshold_(conf.getParameter<double>("minPt_multicluster")),
+    multiclusterAlgoType_(conf.getParameter<string>("type_multicluster")),    
+    nBinsRHisto_(conf.getParameter<unsigned>("nBins_R_histo_multicluster")),
+    nBinsPhiHisto_(conf.getParameter<unsigned>("nBins_Phi_histo_multicluster")),
+    binsSumsHisto_(conf.getParameter< std::vector<unsigned> >("binSumsHisto")),
+    histoThreshold_(conf.getParameter<double>("threshold_histo_multicluster"))
+{    
+    edm::LogInfo("HGCalMulticlusterParameters") << "Multicluster dR: " << dr_;  
+    edm::LogInfo("HGCalMulticlusterParameters") << "Multicluster minimum transverse-momentum: " << ptC3dThreshold_;
+    edm::LogInfo("HGCalMulticlusterParameters") << "Multicluster number of R-bins for the histo algorithm: " << nBinsRHisto_<<endl;
+    edm::LogInfo("HGCalMulticlusterParameters") << "Multicluster number of Phi-bins for the histo algorithm: " << nBinsPhiHisto_<<endl;
+    edm::LogInfo("HGCalMulticlusterParameters") << "Multicluster MIPT threshold for histo threshold algorithm: " << histoThreshold_<<endl;
+    edm::LogInfo("HGCalMulticlusterParameters") << "Multicluster type of multiclustering algortihm: " << multiclusterAlgoType_;
+    id_.reset( HGCalTriggerClusterIdentificationFactory::get()->create("HGCalTriggerClusterIdentificationBDT") );
+    id_->initialize(conf.getParameter<edm::ParameterSet>("EGIdentification"));
+    if(multiclusterAlgoType_.find("Histo")!=std::string::npos && nBinsRHisto_!=binsSumsHisto_.size()) throw cms::Exception("Inconsistent nBins_R_histo_multicluster and binSumsHisto size in HGCalMulticlustering");
+}
+
+
+
+
+
+float HGCalMulticlusteringHistoImpl::dR( const l1t::HGCalCluster & clu,
+					 const GlobalPoint & seed) const
+{
+
+    Basic3DVector<float> seed_3dv( seed );
+    GlobalPoint seed_proj( seed_3dv / seed.z() );
+    return (seed_proj - clu.centreProj() ).mag();
+
+}
+
+
+
+
+
+HGCalMulticlusteringHistoImpl::Histogram HGCalMulticlusteringHistoImpl::fillHistoClusters( const std::vector<edm::Ptr<l1t::HGCalCluster>> & clustersPtrs ){
+
+
+    Histogram histoClusters; //key[0] = z.side(), key[1] = bin_R, key[2] = bin_phi
+
+    for(auto & clu : clustersPtrs){
+
+        float ROverZ = sqrt( pow(clu->centreProj().x(),2) + pow(clu->centreProj().y(),2) );
+        int bin_R = int( (ROverZ-kROverZMin_) * nBinsRHisto_ / (kROverZMax_-kROverZMin_) );
+        int bin_phi = int( (reco::reduceRange(clu->phi())+M_PI) * nBinsPhiHisto_ / (2*M_PI) );
+
+        std::array<int,3> key = { { clu->zside(), bin_R, bin_phi } };
+        histoClusters[key]+=clu->mipPt();
```

Comment:
In all other places the histos are accessed without temporary key, the same could be done here:
```cpp
histoClusters[{ { clu->zside(), bin_R, bin_phi } }] += clu->mipPt();
```